### PR TITLE
Warn about inconsistent indentation

### PR DESCRIPTION
**Pull request by phadej**
This is an effect of using `indentOfAtLeast` method: any indentation greater than current offset is fine.

That behavior is desirable to parsing multiline field contents, but it is a bit surprising for fields in sections, which we expect to be aligned.

Such insonsistency seems to be always a mistake, and it's easy to fix once a machine points it out.

### CODE REVIEW

File: Cabal-tests/tests/ParserTests/regressions/decreasing-indentation.check [hunk 1]
```diff
@@ -0,0 +1,16 @@
+decreasing-indentation.cabal:40:3: Inconsistent indentation at lines 40, 41, 42, 43, 44, 50, 52, 57, 58, 70, 71
+No 'main-is' field found for executable warnings
+Deprecated extensions: 'PatternSignatures'. Instead of 'PatternSignatures' use 'ScopedTypeVariables'.
+'ghc-options: -O2' is rarely needed. Check that it is giving a real benefit and not just imposing longer compile times on your users.
+These packages miss upper bounds:
```
Comment (ffaf1) [inline Cabal-tests/tests/ParserTests/regressions/decreasing-indentation.check:5]: I would suggest to fix those missing-upper-bounds warnings (actually: any warning that is not “decreasing indentation”). Makes the golden test clearer and more robust.

File: Cabal-tests/tests/ParserTests/regressions/decreasing-indentation.check [hunk 2]
```diff
@@ -0,0 +1,16 @@
+decreasing-indentation.cabal:40:3: Inconsistent indentation at lines 40, 41, 42, 43, 44, 50, 52, 57, 58, 70, 71
```
Comment (ffaf1) [inline Cabal-tests/tests/ParserTests/regressions/decreasing-indentation.check:1]: Maybe `40-44, 50, 52, 57-58, 70-71` would be clearer?

Comment (ffaf1) [inline Cabal-tests/tests/ParserTests/regressions/decreasing-indentation.check:1]: How would I fix 70-71? By having `if` ⌫ 3?

File: Cabal-tests/tests/ParserTests/regressions/decreasing-indentation.check (hunk 1 above)
Comment (phadej) [reply Cabal-tests/tests/ParserTests/regressions/decreasing-indentation.check:5]: The test file vanilla `RSA-1.0.0` file (with some extra additions at the end). I'd rather leave it as is, if this is not a hard requirement for PR acceptance.

File: Cabal-tests/tests/ParserTests/regressions/decreasing-indentation.check (hunk 2 above)
Comment (phadej) [reply Cabal-tests/tests/ParserTests/regressions/decreasing-indentation.check:1]: Yes

Comment (phadej) [reply Cabal-tests/tests/ParserTests/regressions/decreasing-indentation.check]: I changed the code to report the warning only at lines where indentation jumps, so now it is `Indentation jumps at lines 40, 50, 57, 58, 70` which is more to the point.

**Comment by phadej** [comment]
I had to adjust lexer slightly as BOM and NBSP where throwing source locations a bit off causing cascading indentation warnings.

File: Cabal-tests/tests/ParserTests/regressions/decreasing-indentation.check (hunk 1 above)
Comment (ulysses4ever) [reply Cabal-tests/tests/ParserTests/regressions/decreasing-indentation.check:5]: I agree that isolating the warning we test for is desirable. One reason is that if other warnings change in the future, we won’t have to touch this test result.

Why is it good (as seemingly implied) to have a vanilla cabal file of a random package in Cabal test suite? I don’t see any reason to strive for such files.

Comment (phadej) [reply Cabal-tests/tests/ParserTests/regressions/decreasing-indentation.check]: Done

File: changelog.d/inconsistent-indentation [hunk 3]
```diff
@@ -0,0 +1,12 @@
+synopsis: Warn about inconsistent indentation
+packages: Cabal-syntax
+prs: #8975
+
+description:
+    Make readFields warn about inconsistent indentation.
```
Comment (ulysses4ever) [inline changelog.d/inconsistent-indentation:6]: Could the changelog start in a more approachable manner explaining the overall effect of the change for a user (e.g. "if you have this and that in your cabal file, you'll be warned now" or some such...)? Starting referencing API from get go feels intimidating. Changelogs go to release notes and as such will probably be read by many people, most of whom are unfamiliar with Cabal code base and API.

**Review by ulysses4ever — COMMENTED** [review]
> I had to adjust lexer slightly as BOM and NBSP where throwing source locations a bit off causing cascading indentation warnings.

This sounds a bit worrying. Have you seen the discussion at #8896? Don't we step on each others toes here?

At the very least, the Lexer change sounds like something that should go into a separate commit.

File: changelog.d/inconsistent-indentation (hunk 3 above)
Comment (phadej) [reply changelog.d/inconsistent-indentation]: > > I had to adjust lexer slightly as BOM and NBSP where throwing source locations a bit off causing cascading indentation warnings.
> 
> This sounds a bit worrying. Have you seen the discussion at #8896? Don't we step on each others toes here?
> 
> At the very least, the Lexer change sounds like something that should go into a separate commit.

I haven't. Can you summarize?

I split that change into separate commit.

Also I modified the changelog entry. Please suggest concrete edits if you want it to be changed further.

Comment (ulysses4ever) [reply changelog.d/inconsistent-indentation]: Thank you.

@hsyl20, as the last person to touch Lexer.x, could you perhaps share your view on the proposed change and say if it looks reasonable to you?

Comment (ulysses4ever) [reply changelog.d/inconsistent-indentation]: Thank you!

**Review by ulysses4ever — APPROVED** [review]

**Comment by phadej** [comment]
ping

**Review by Mikolaj — APPROVED** [review]
Thanks a lot. Please kindly set one of the merge labels.

**Comment by phadej** [comment]
I don't have (and don't want) write permissions to this repository.

**Comment by ulysses4ever** [comment]
@phadej no worries, I did the labeling. Thanks again! Btw, #8983 has one commit in common with this PR, so maybe it'll need a rebase at some point...

**Comment by phadej** [comment]
Why isn't this merged? There isn't conflict?

**Comment by ulysses4ever** [comment]
@mergify rebase

**Comment by ulysses4ever** [comment]
@phadej on it. Sorry for the delay.

**Comment by ulysses4ever** [comment]
@mergify refresh

**Comment by Mikolaj** [comment]
@mergify refresh

**Comment by Mikolaj** [comment]
Oh dear.

**Comment by Mikolaj** [comment]
@phadej: we have technical problems; is it fine if I squash the PR when merging? If not, we'd probably need to open a new PR and move it there, but it's work and an indirection.

**Comment by ulysses4ever** [comment]
@Mikolaj Oleg was [very clear about it](https://github.com/haskell/cabal/pull/8981#issuecomment-1597278210): he won't allow updates. We'll have to manually copy his branches to our repository and open new PRs using those copies. I am afraid it'll be hard to find volunteers to do this monkey job.

**Comment by Mikolaj** [comment]
@ulysses4ever: I thought the merge+squash button would be a cheap way out. Wouldn't it work? In any case, @ffaf1did it in a more principled way, so let's merge his PR.

**Comment by ulysses4ever** [comment]
@Mikolaj unless I'm missing something, Mergify is completely strangled by not being able to update the branch. Perhaps, you could work around Mergify by clicking GitHub buttons for squash and merge, see the screenshot.
Screenshot_20230620-105247

But I find it a bit dangerous to apply both merge approaches (GH and Mergify) on one repository concurrently.

**Comment by Mikolaj** [comment]
Yes, the button is what I meant. If the button works, then mergify does not cut in. Unless I'm missing something. However we lose the separate commits.

**Comment by Mikolaj** [comment]
Huh, I'd hope git would cope fine and then mergify or github trying to merge something absurd would be rebuked by git. OTOH, if they promptly retry with enough `reset --hard` to convince git, I'm afraid your vision may be right. :[
